README.pod
=head1 NAME

Boost::Geometry::Utils - Bindings for the Boost Geometry library

=head1 VERSION

version 0.06

=head1 SYNOPSIS

    use Boost::Geometry::Utils qw(polygon linestring polygon_linestring_intersection);
    
    my $square = [  # ccw
        [10, 10],
        [20, 10],
        [20, 20],
        [10, 20],
    ];
    my $hole_in_square = [  # cw
        [14, 14],
        [14, 16],
        [16, 16],
        [16, 14],
    ];
    my $polygon = polygon($square, $hole_in_square);
    my $linestring = linestring([ [5, 15], [30, 15] ]);
    
    my $intersection = polygon_linestring_intersection($polygon, $linestring);
    
    # $intersection is:
    # [
    #     [ [10, 15], [14, 15] ],
    #     [ [16, 15], [20, 15] ],
    # ]

=head1 ABSTRACT

This module provides bindings to perform some geometric operations using
the Boost Geometry library. It does not aim at providing full bindings
for such library, and that's why I left the I<Boost::Geometry> namespace
free. I'm unsure about the optimal architectural for providing full 
bindings, but I'm interested in such a project -- so, if you have ideas
please get in touch with me.

B<Warning:> the API could change in the future.

=head1 METHODS

=head2 polygon_linestring_intersection

Performs an intersection between the supplied polygon and linestring,
and returns an arrayref of linestrings (represented as arrayrefs of
points).
Note that such an intersection is also called I<clipping>.

=head2 polygon_multi_linestring_intersection

Same as I<polygon_linestring_intersection> but it accepts a multilinestring
object to perform multiple clippings in a single batch.

=head2 multi_polygon_multi_linestring_intersection

Same as I<polygon_multi_linestring_intersection> but it accepts a multipolygon
object to perform multiple clippings in a single batch.

=head2 polygon_to_wkt

Converts one or more arrayref(s) of points to a WKT representation of
a polygon (with holes).

=head2 linestring_to_wkt

Converts an arrayref of points to a WKT representation of a multilinestring.

=head2 wkt_to_multilinestring

Parses a MULTILINESTRING back to a Perl data structure.

=head2 linestring_simplify

Accepts an arrayref of points representing a linestring and a numeric tolerance 
and returns an arrayref of points representing the simplified linestring.

=head2 multi_linestring_simplify

Accepts an arrayref of arrayrefs of points representing a multilinestring and a 
numeric tolerance and returns an arrayref of arrayrefs of points representing 
the simplified linestrings.

=head2 point_covered_by_polygon

Accepts a point and an arrayref of points representing a polygon and returns true 
or false according to the 'cover_by' strategy.

=head2 point_within_polygon

Accepts a point and an arrayref of points representing a polygon and returns true 
or false according to the 'within' strategy.

=head2 linestring_length

Returns length of a linestring.

=head2 polygon_centroid

Returns the centroid point of a given polygon.

=head2 linestring_centroid

Returns the centroid point of a given linestring.

=head2 multi_linestring_centroid

Returns the centroid point of a given multi_linestring.

=head2 correct_polygon

Corrects the orientation(s) of the given polygon.

=head2 correct_multi_polygon

Corrects the orientation(s) of the given multi_polygon.

=head1 ACKNOWLEDGEMENTS

Thanks to mauke and mst (Matt S. Trout (cpan:MSTROUT) <mst@shadowcat.co.uk>)
for their valuable help in getting this to compile under Windows (MinGW) too.
Thanks to Mark Hindness for his work on data types conversion.

=head1 AUTHOR

Alessandro Ranellucci <aar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Alessandro Ranellucci.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



CHANGES
==================================================
Changes from 2012-03-31 00:00:00 +0000 to present.
==================================================

    - Added point_within_multi_polygon()
    - Added point_covered_by_multi_polygon()

-----------------------------------------
version 0.14 at 2013-06-16
-----------------------------------------

    - Fix segfault on reading empty multilinestrings

-----------------------------------------
version 0.13 at 2013-05-20
-----------------------------------------

    - Remove a possible compilation error

-----------------------------------------
version 0.12 at 2013-05-19
-----------------------------------------

    - Fix compilation on some GCC

-----------------------------------------
version 0.11 at 2013-05-19
-----------------------------------------

    - Added polygon_area()
    - Fixed more memory leaks

-----------------------------------------
version 0.10 at 2013-05-18
-----------------------------------------

    - Added multi_linestring_multi_polygon_difference()
    - Fixed memory leak in reading multi_linestring input
    - Fixed memory leak in reading multi_polygon input

-----------------------------------------
version 0.09 at 2013-04-11
-----------------------------------------

    - fix compilation on Windows after recent Boost update

-----------------------------------------
version 0.08 at 2013-04-08
-----------------------------------------

    - Updated Boost to 1.53
    - Revert to double coordinates instead of integer because Boost computes
      wrong intersection with integers

-----------------------------------------
version 0.07 at 2013-03-31 17:03:07 +0000
-----------------------------------------

  Change: cbbf35be3e139a5b21773fa2cf0d9b545dec0d44
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-31 11:45:00 +0000

    Add hints for Pod::Coverage 

  Change: 0defa9741671173d2d5395be2343b0e36e266a4e
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-15 15:47:04 +0000

    New correct_polygon() and correct_multi_polygon() 

  Change: 2b190280466cc54d44f7f1ba7b1a133dcfaeaa07
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-15 15:10:44 +0000

    Added multi_polygon_multi_linestring_intersection() 

  Change: b9d2e24cabae84d503759a9b9c096f488064be2c
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-14 17:50:01 +0000

    Added multi_linestring_centroi() 

  Change: 66e6cf92fa9594d74e6edf433a7adcd80b73289b
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-14 17:40:33 +0000

    Added linestring_centroid() 

  Change: 1a5fc63410c5433cfced29f909426e0bf677828b
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-14 17:37:27 +0000

    Added polygon_centroid() 

  Change: dd48d474f1b4b75a24f96589bad1965c81aa2339
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-14 17:18:19 +0000

    Added linestring_length 

  Change: 04167004ef25d6fe22c6bfcc6e0463a35c7d06da
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-14 17:05:26 +0000

    Added multi_linestring_simplify and fixed some files 

  Change: db7138df02bf1e5078f7ebe9d68a196603bca3e2
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-03-14 16:50:42 +0000

    Merge branch 'master' of github.com:alexrj/Boost-Geometry-Utils 

  Change: 29af743cedd602af65deecf056b59ac0c909282a
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-28 00:25:03 +0000

    Update POD 

-----------------------------------------
version 0.06 at 2013-02-01 23:04:44 +0000
-----------------------------------------

  Change: e2b880930b6f1ca3015d34ee0099f1c629f103ed
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-02-02 00:04:44 +0000

    Update Boost to 1.52 

  Change: 58736a034fb398d7e4f6a00b975a14ce2d122d7a
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 23:58:02 +0000

    Fixed arrayref to polygon conversion and added unit test 

  Change: 885c60668125cd281b7783baf7dc69951e0d0f2e
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 16:30:38 +0000

    Finished polygon2perl() and added unit test for conversion roundtrip 

  Change: 1df409ec73fb4de93db234f9634399c64aba0314
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 15:08:11 +0000

    Update copyright year 

  Change: 0a0b9442fb84af318e888a0064da6f54474fc9ae
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 15:07:56 +0000

    Added a couple conversion methods to help debugging 

  Change: c3fc121bf240a9b15b69d356b4ffc8088ed52609
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-11-01 17:18:17 +0000

    Use integers instead of doubles and fix test to use integers. 

  Change: 71ce53ba110bea82bb32c63e8aef82719bcbea80
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-10 22:28:25 +0000

    Convert polygon() and linestring() functions to use non-wkt
    conversion. 

  Change: f95918786093c58de80ab62c0f47355c668fb48e
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-08 11:41:27 +0000

    Attempt to fix/workaround leaking memory. 

  Change: fec904e6630837aef63260b5bdde034b6d570ae5
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-08 08:30:55 +0000

    Remove pointless reference to boost::geometry::append. 

  Change: e917ff4dbc3eeaa3e2e51cacb0dad649712310d1
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-08 07:50:09 +0000

    Add point_covered_by_polygon and add tests to cover "edge" cases. 

  Change: 259eed2f7952443be651f87d50772181a363c105
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 23:40:35 +0000

    Add point_within_polygon and linestring_simplify functions and tests.

    There is a bug in point_within_polygon triggered by:

    point_within_polygon($point_in_hole, [$hole_in_square,[]])

    which I've avoided by changing the test. This needs fixing. 


src/myinit.h
#undef read
#undef bind
#undef times
#undef open
#undef seekdir
#undef setbuf
#undef abort
#undef close
#undef select
#include <boost/geometry.hpp>
#include <boost/geometry/geometries/point_xy.hpp>
#include <boost/geometry/geometries/polygon.hpp>
#include <boost/geometry/geometries/linestring.hpp>
#include <boost/geometry/geometries/ring.hpp>
#include <boost/geometry/multi/geometries/multi_linestring.hpp>
#include <boost/geometry/multi/geometries/multi_polygon.hpp>
#include <boost/geometry/io/wkt/wkt.hpp>

typedef boost::geometry::model::d2::point_xy<double> point_xy;
typedef boost::geometry::model::polygon<point_xy,false,false> polygon;
typedef boost::geometry::model::multi_polygon<polygon> multi_polygon;
typedef boost::geometry::model::linestring<point_xy> linestring;
typedef boost::geometry::model::multi_linestring<linestring> multi_linestring;
typedef boost::geometry::model::ring<point_xy,false,false> ring;

// Support old API
typedef boost::geometry::model::polygon<point_xy,false,false> opolygon;
typedef boost::geometry::model::multi_polygon<polygon> omultipolygon;
typedef boost::geometry::model::multi_linestring<linestring> omultilinestring;


// for av_fetch, use SvIV() 
// this library then supports 64 bit ints.
#define av_fetch_x(AV) SvNV(*av_fetch(AV, 0, 0))
#define av_fetch_y(AV) SvNV(*av_fetch(AV, 1, 0))

// for av_store, use newSViv()
#define av_store_point_xy(AV, X, Y)              \
  av_store(AV, 0, newSVnv(X));                   \
  av_store(AV, 1, newSVnv(Y))

#define av_fetch_point_xy(AV)                    \
  make<point_xy>(av_fetch_x(AV), av_fetch_y(AV))

#include "poly2av.h"
#include "mline2av.h"
#include "point2av.h"
#include "line2av.h"


src/line2av.h
#ifndef bgu_line2av_h_
#define bgu_line2av_h_

#include "myinit.h"
#include <boost/geometry/algorithms/num_points.hpp>

SV*
linestring2perl(pTHX_ const linestring& ls)
{
  AV* av = newAV();
  const unsigned int line_len = boost::geometry::num_points(ls);
  av_extend(av, line_len-1);

  for (int i = 0; i < line_len; i++) {
    AV* pointav = newAV();
    av_store(av, i, newRV_noinc((SV*)pointav));
    av_fill(pointav, 1);
    av_store_point_xy(pointav, ls[i].x(), ls[i].y());
  }
    
  return (SV*)newRV_noinc((SV*)av);
}

linestring*
perl2linestring(pTHX_ AV* theAv)
{
    using boost::geometry::make;

  const unsigned int len = av_len(theAv)+1;
  SV** elem;
  AV* innerav;
  elem = av_fetch(theAv, 0, 0);
  if (!SvROK(*elem)
      || SvTYPE(SvRV(*elem)) != SVt_PVAV
      || av_len((AV*)SvRV(*elem)) < 1) {
    return NULL;
  }
  linestring* retval = new linestring();
  
  for (unsigned int i = 0; i < len; i++) {
    elem = av_fetch(theAv, i, 0);
    if (!SvROK(*elem)
        || SvTYPE(SvRV(*elem)) != SVt_PVAV
        || av_len((AV*)SvRV(*elem)) < 1)
    {
      delete retval;
      return NULL;
    }
    innerav = (AV*)SvRV(*elem);
    retval->push_back(av_fetch_point_xy(innerav));
  }
  return retval;
}

#endif


src/poly2av.h
#ifndef bgu_poly2av_h_
#define bgu_poly2av_h_

#include <iostream>
#include "myinit.h"
#include <boost/geometry/algorithms/num_points.hpp>

void add_ring_perl(AV* polygon_av, ring& theRing)
{
    AV* ring_av = newAV();
    AV* point_av;
    
    // number of points in the ring
    const unsigned int len = boost::geometry::num_points(theRing);
    av_extend(ring_av, len-1);
    //fprintf(stderr, "Points in ring: len=%d\n", len);
  
    for(unsigned int i = 0; i < len; i++) {
        point_av = newAV();
        av_store(ring_av, i, newRV_noinc((SV*)point_av));
        av_fill(point_av, 1);
        av_store_point_xy(point_av, theRing[i].x(), theRing[i].y());
    }
    
    av_push(polygon_av, newRV_noinc((SV*)ring_av));
}

SV*
polygon2perl(pTHX_ const polygon& poly)
{
    AV* av = newAV();
    
    ring my_ring = poly.outer();
    add_ring_perl(av, my_ring);
    
    std::vector<ring>::size_type sz = poly.inners().size();
    for (unsigned i = 0; i < sz; i++) {
        my_ring = poly.inners()[i];
        add_ring_perl(av, my_ring);
    }
    
    return (SV*)newRV_noinc((SV*)av);
}

SV*
multi_polygon2perl(pTHX_ const multi_polygon& multi_poly)
{
    AV* av = newAV();
    
    // number of polygons in this multi_polygon
    std::size_t poly_count = boost::geometry::num_geometries(multi_poly);
    
    for (unsigned i = 0; i < poly_count; i++) {
        polygon poly = multi_poly[i];
        AV* polyav = newAV();
        
        // add contour
        ring my_ring = poly.outer();
        add_ring_perl(polyav, my_ring);
        
        // add holes
        std::vector<ring>::size_type sz = poly.inners().size();
        for (unsigned j= 0; j < sz; j++) {
            my_ring = poly.inners()[j];
            add_ring_perl(polyav, my_ring);
        }
        
        av_push(av, newRV_noinc((SV*)polyav));
    }
    
    return (SV*)newRV_noinc((SV*)av);
}

int add_ring(AV* theAv, polygon& poly, const int ring_index)
{
    using boost::geometry::append;
    using boost::geometry::make;

  const unsigned int len = av_len(theAv)+1;
  SV** elem;
  AV* innerav;
  for (unsigned int i = 0; i < len; i++) {
    elem = av_fetch(theAv, i, 0);
    if (!SvROK(*elem)
        || SvTYPE(SvRV(*elem)) != SVt_PVAV
        || av_len((AV*)SvRV(*elem)) != 1)  // require two coordinates per point
    {
      return 0;
    }
    innerav = (AV*)SvRV(*elem);
    append(poly, av_fetch_point_xy(innerav), ring_index);
  }
  return 1;
}

polygon*
perl2polygon(pTHX_ AV* theAv)
{
    using boost::geometry::interior_rings;
    
    // read number of input rings
    const unsigned int len = av_len(theAv)+1;
    
    // initialize resulting polygon
    polygon* retval = new polygon();
    
    // create interior rings (append doesn't create them automatically)
    interior_rings(*retval).resize(len-1);
  
    SV** elem;
    for (unsigned int i = 0; i < len; i++) {
        // validate input data
        elem = av_fetch(theAv, i, 0);
        if (!SvROK(*elem)
            || SvTYPE(SvRV(*elem)) != SVt_PVAV
            || av_len((AV*)SvRV(*elem)) < 1)  // require at least two points in ring
        {
            delete retval;
            return NULL;
        }
    
        // append point to ring i-1 (-1 is outer ring)
        if (!add_ring((AV*)SvRV(*elem), *retval, i-1)) {
            delete retval;
            return NULL;
        }
    }
    return retval;
}

multi_polygon*
perl2multi_polygon(pTHX_ AV* theAv)
{
    // initialize resulting multi_polygon
    multi_polygon* retval = new multi_polygon();
  
    SV** elem;
    const unsigned int len = av_len(theAv)+1;
    for (unsigned int i = 0; i < len; i++) {
        // validate input data
        elem = av_fetch(theAv, i, 0);
        if (!SvROK(*elem)
            || SvTYPE(SvRV(*elem)) != SVt_PVAV
            || av_len((AV*)SvRV(*elem)) < 0)  // require at least one polygon
        {
            delete retval;
            return NULL;
        }
        
        polygon* poly = perl2polygon(aTHX_ (AV*)SvRV(*elem));
        retval->push_back(*poly);
    }
    return retval;
}

#endif


src/mline2av.h
#ifndef bgu_mline2av_h_
#define bgu_mline2av_h_

#include "myinit.h"
#include <boost/geometry/algorithms/num_points.hpp>

SV*
multi_linestring2perl(pTHX_ const multi_linestring& mls)
{
  AV* av = newAV();
  const unsigned int size = mls.size();
  av_extend(av, size-1);

  for (int i = 0; i < size; i++) {
    AV* lineav = newAV();
    linestring ls = mls[i];
    av_store(av, i, newRV_noinc((SV*)lineav));
    av_fill(lineav, 1);
    const unsigned int line_len = boost::geometry::num_points(ls);
    for (int j = 0; j < line_len; j++) {
      AV* pointav = newAV();
      av_store(lineav, j, newRV_noinc((SV*)pointav));
      av_fill(pointav, 1);
      av_store_point_xy(pointav, ls[j].x(), ls[j].y());
    }
  }
    
  return (SV*)newRV_noinc((SV*)av);
}

void add_line(AV* theAv, multi_linestring* mls)
{
    using boost::geometry::make;

  const unsigned int len = av_len(theAv)+1;
  SV** elem;
  AV* innerav;
  linestring ls;
  for (unsigned int i = 0; i < len; i++) {
    elem = av_fetch(theAv, i, 0);
    if (!SvROK(*elem)
        || SvTYPE(SvRV(*elem)) != SVt_PVAV
        || av_len((AV*)SvRV(*elem)) < 1)
    {
      return;
    }
    innerav = (AV*)SvRV(*elem);
    ls.push_back(av_fetch_point_xy(innerav));
  }
  mls->push_back(ls);
}

multi_linestring*
perl2multi_linestring(pTHX_ AV* theAv)
{
    
  const unsigned int len = av_len(theAv)+1;
  SV** elem;
  AV* innerav;
  elem = av_fetch(theAv, 0, 0);
  if (!SvROK(*elem)
      || SvTYPE(SvRV(*elem)) != SVt_PVAV
      || av_len((AV*)SvRV(*elem)) < 1) {
    return NULL;
  }
  multi_linestring* retval = new multi_linestring();
  
  for (unsigned int i = 0; i < len; i++) {
    elem = av_fetch(theAv, i, 0);
    if (!SvROK(*elem)
        || SvTYPE(SvRV(*elem)) != SVt_PVAV
        || av_len((AV*)SvRV(*elem)) < 1)
    {
      delete retval;
      return NULL;
    }
    add_line((AV*)SvRV(*elem), retval);
  }
  return retval;
}

#endif


lib/Boost/Geometry/Utils.pm
package Boost::Geometry::Utils;
{
  $Boost::Geometry::Utils::VERSION = '0.14';
}
# ABSTRACT: Bindings for the Boost Geometry library
use strict;
use warnings;

require Exporter;
our @ISA = qw(Exporter);

use XSLoader;
XSLoader::load('Boost::Geometry::Utils', $Boost::Geometry::Utils::VERSION);

our @EXPORT_OK = qw(polygon_to_wkt linestring_to_wkt wkt_to_multilinestring
    polygon linestring polygon_linestring_intersection
    polygon_multi_linestring_intersection multi_polygon_multi_linestring_intersection
    point_within_polygon point_covered_by_polygon
    point_within_multi_polygon point_covered_by_multi_polygon
    linestring_simplify
    multi_linestring_simplify linestring_length polygon_centroid
    linestring_centroid multi_linestring_centroid multi_polygon
    correct_polygon correct_multi_polygon multi_linestring_multi_polygon_difference
    polygon_medial_axis polygon_area);

sub polygon_to_wkt {
    sprintf 'POLYGON(%s)', join ',', map { sprintf '(%s)', join ',', map { join ' ', @$_ } @$_ } @_;
}

sub linestring_to_wkt {
    sprintf "MULTILINESTRING(%s)", join ',', map { sprintf '(%s)', join ',', map { join ' ', @$_ } @$_ } @_;
}

sub wkt_to_multilinestring {
    return [] if $_[0] eq 'MULTILINESTRING()';
    $_[0] =~ s/^MULTILINESTRING\(\(//;
    $_[0] =~ s/\)\)$//;
    [ map [ map [ split / / ], split /,/ ], split /\),\(/, $_[0] ];
}

sub polygon {
    _polygon(\@_);
}

sub multi_polygon {
    _multi_polygon(\@_);
}

sub linestring {
    _multi_linestring(\@_)
}

sub multi_linestring {
    _multi_linestring(\@_)
}

1;


__END__
=pod

=head1 NAME

Boost::Geometry::Utils - Bindings for the Boost Geometry library

=head1 VERSION

version 0.07

=head1 SYNOPSIS

    use Boost::Geometry::Utils qw(polygon linestring polygon_linestring_intersection);
    
    my $square = [  # ccw
        [10, 10],
        [20, 10],
        [20, 20],
        [10, 20],
    ];
    my $hole_in_square = [  # cw
        [14, 14],
        [14, 16],
        [16, 16],
        [16, 14],
    ];
    my $polygon = polygon($square, $hole_in_square);
    my $linestring = linestring([ [5, 15], [30, 15] ]);
    
    my $intersection = polygon_linestring_intersection($polygon, $linestring);
    
    # $intersection is:
    # [
    #     [ [10, 15], [14, 15] ],
    #     [ [16, 15], [20, 15] ],
    # ]

=head1 ABSTRACT

This module provides bindings to perform some geometric operations using
the Boost Geometry library. It does not aim at providing full bindings
for such library, and that's why I left the I<Boost::Geometry> namespace
free. I'm unsure about the optimal architectural for providing full 
bindings, but I'm interested in such a project -- so, if you have ideas
please get in touch with me.

B<Warning:> the API could change in the future.

=head1 METHODS

=head2 polygon_linestring_intersection

Performs an intersection between the supplied polygon and linestring,
and returns an arrayref of linestrings (represented as arrayrefs of
points).
Note that such an intersection is also called I<clipping>.

=head2 polygon_multi_linestring_intersection

Same as I<polygon_linestring_intersection> but it accepts a multilinestring
object to perform multiple clippings in a single batch.

=head2 multi_polygon_multi_linestring_intersection

Same as I<polygon_multi_linestring_intersection> but it accepts a multipolygon
object to perform multiple clippings in a single batch.

=head2 multi_linestring_multi_polygon_difference

Performs a difference between the supplied multilinestring and the supplied
multipolygon. It returns a multilinestring object.

=head2 polygon_to_wkt

Converts one or more arrayref(s) of points to a WKT representation of
a polygon (with holes).

=head2 linestring_to_wkt

Converts an arrayref of points to a WKT representation of a multilinestring.

=head2 wkt_to_multilinestring

Parses a MULTILINESTRING back to a Perl data structure.

=head2 linestring_simplify

Accepts an arrayref of points representing a linestring and a numeric tolerance 
and returns an arrayref of points representing the simplified linestring.

=head2 multi_linestring_simplify

Accepts an arrayref of arrayrefs of points representing a multilinestring and a 
numeric tolerance and returns an arrayref of arrayrefs of points representing 
the simplified linestrings.

=head2 point_covered_by_polygon

Accepts a point and an arrayref of points representing a polygon and returns true 
or false according to the 'cover_by' strategy.

=head2 point_covered_by_multi_polygon

Same as above but accepts a multipolygon arrayref.

=head2 point_within_polygon

Accepts a point and an arrayref of points representing a polygon and returns true 
or false according to the 'within' strategy.

=head2 point_within_multi_polygon

Same as above but accepts a multipolygon arrayref.

=head2 linestring_length

Returns length of a linestring.

=head2 polygon_centroid

Returns the centroid point of a given polygon.

=head2 linestring_centroid

Returns the centroid point of a given linestring.

=head2 multi_linestring_centroid

Returns the centroid point of a given multi_linestring.

=head2 correct_polygon

Corrects the orientation(s) of the given polygon.

=head2 correct_multi_polygon

Corrects the orientation(s) of the given multi_polygon.

=head2 polygon_area

Returns the area of the given polygon.

=for Pod::Coverage linestring multi_linestring multi_polygon polygon

=head1 ACKNOWLEDGEMENTS

Thanks to mauke and mst (Matt S. Trout (cpan:MSTROUT) <mst@shadowcat.co.uk>)
for their valuable help in getting this to compile under Windows (MinGW) too.
Thanks to Mark Hindness for his work on data types conversion.

=head1 AUTHOR

Alessandro Ranellucci <aar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Alessandro Ranellucci.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

